### PR TITLE
First pass at crypto callback docs

### PR DESCRIPTION
**Pull request by bigbrett**

### CODE REVIEW

**Comment by dgarske** [comment]
GitHub CI seems to be failing with `1.952 cat: api/md/tpm2_8h.md: No such file or directory`?

File: wolfSSL/src/chapter06.md [hunk 1]
```diff
@@ -185,3 +185,158 @@ These contexts may be pointers to any user-specified context, which will then in
...
+### Handling the request
+
+The data structure for each type of request generally contains a pointer and associated size for input and output data. Depending on the request it may include additional data including cryptographic keys, nonces, or further configuration. The crypto callback should operate on the input data and write relevant output data back to the appropriate variant of the algo type union. Writing data to a union variant that does not correspond to the algorithm type in question (e.g. using `info->cipher` when `info->algo_type == WC_ALGO_TYPE_RNG` is a memory error and can result in undefined behavior.
+
+
+### Examples
+
+Full examples of crypto callbacks can be found in the following locations
+
+    VaultIC Crypto Callbacks: https://github.com/wolfSSL/wolfssl-examples/blob/master/ccb_vaultic/ccb_vaultic.c
```
Comment (dgarske) [inline wolfSSL/src/chapter06.md:338]: Please use hyperlink markdown formatting. I also fixed the wolfTPM link

Full examples of crypto callbacks can be found in the following locations:

* [VaultIC Crypto Callbacks](https://github.com/wolfSSL/wolfssl-examples/blob/master/ccb_vaultic/ccb_vaultic.c)
* [STSAFE-A100 ECC Crypto Callbacks](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/src/port/st/stsafe.c)
* [TPM 2.0 wolfTPM Crypto Callbacks](https://github.com/wolfSSL/wolfTPM/blob/master/src/tpm2_cryptocb.c)
* [Generic wolfCrypt tests](https://github.com/wolfSSL/wolfssl/blob/master/wolfcrypt/test/test.c)

File: wolfSSL/src/chapter06.md [hunk 2]
```diff
@@ -185,3 +185,158 @@ These contexts may be pointers to any user-specified context, which will then in
...
+3. Pass the `devId` as an argument to a wolfCrypt API function
+4. (TLS only): associate the `devId` with a wolfSSL context
+
+
+#### 1. Compile wolfSSL with crypto callback support
+Support for crypto callbacks is enabled in wolfSSL via the `–enable-cryptocb` configure option, or `#define WOLF_CRYPTO_CB`.
+
+#### 2. Register your callback and unique devID with wolfCrypt
+
+To use crypto callbacks, the user must register a top-level cryptographic callback function with wolfCrypt and provide a device ID (`devId`), which is an integer ID that uniquely identifies this particular callback function to wolfCrypt. This allows multiple crypto callbacks to be registered with wolfCrypt at once. The appropriate callback function will then be invoked internally by every wolfCrypt function that takes a `devId` parameter, or by wolfSSL when associated with a `WOLFSSL_CTX` or `WOLFSSL` structure.
```
Comment (dgarske) [inline wolfSSL/src/chapter06.md:209]: Please add comment about `INVALID_DEVID` and it being `-2`.

File: wolfSSL/src/chapter06.md [hunk 3]
```diff
@@ -185,3 +185,158 @@ These contexts may be pointers to any user-specified context, which will then in
...
+
+### Writing your crypto callback
+
+When a crypto callback is invoked, it will need to know what cryptographic operation has been requested, as well as the location of the input and output data and how to communicate success/failure back to wolfCrypt. This information is relayed to the callback through the `wc_CryptoInfo* info` argument.
+
+At a high level, a crypto callback should determine if it supports the requested operation, act on the input data, update the `wc_CryptoInfo *info` argument with any relevant output data, and return a valid wolfCrypt error code.
+
+#### Determining the Type of Request
+
+The `wc_CryptoInfo` structure contains a union of different structures (some of which are unions themselves), each corresponding to a specific type of cryptographic operation. The type of operation to be performed is determined by `info->algo_type`, which should be a variant of `enum wc_AlgoType` defined in `wolfssl/wolfcrypt/types.h`.  Based on this `algo_type`, the appropriate union element within `wc_CryptoInfo` can be accessed to get or set parameters specific to the operation. We can refer to this top-level union as the "algo type union".
```
Comment (dgarske) [inline wolfSSL/src/chapter06.md:252]: Add comment about `HAVE_ANONYMOUS_INLINE_AGGREGATES` if getting error about compiler not allowing anonymous unions.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Comment by dgarske** [comment]
@bigbrett : By the way... Thank you so much for documenting this! Excellent work

File: wolfSSL/src/chapter06.md (hunk 1 above)
Comment (bigbrett) [reply wolfSSL/src/chapter06.md]: done

File: wolfSSL/src/chapter06.md (hunk 2 above)
Comment (bigbrett) [reply wolfSSL/src/chapter06.md:209]: done

**Review by dgarske — APPROVED** [review]

**Pull request merged**